## Introduction: 

In this notebook we gonna perform some statistical analysis on pmf concept ..

i.e. compute the mean and variance for pmf .. 

## Ex. 
Write functions called PmfMean and PmfVar that take a Pmf object and compute the mean and variance. To test these methods, check that they are consistent with the methods Mean and Var in Pmf.py

In [1]:
using DataFrames

We gonna copy the same list of elements as Python code .. just to be able to compare the results .. 

In [3]:
data = [30, 20, 25, 40, 50, 25]
data

6-element Array{Int64,1}:
 30
 20
 25
 40
 50
 25

In [35]:
c = countmap(data)  ## create data dictionary 

Dict{Int64,Int64} with 5 entries:
  20 => 1
  25 => 2
  50 => 1
  40 => 1
  30 => 1

## 1. Calc Mean for PMF:

In [39]:
function PmfMean(c)
    pmf = []
    index = []

    for (k, v) in c
        append!(pmf, v/length(data))
        append!(index, k)       
    end
    
    return sum(index.*pmf)
end

PmfMean (generic function with 1 method)

In [41]:
m = PmfMean(c)

31.666666666666664

- The mean value for PMF generated by Julia matches that one generated in Python as well as the built in function implemented by the uthor of the textbook .. 

- I think the value of 31.667 makes sense .. as the distribution of the values is nearly normal .. and we have not got any outliers hence, it is expected to have a mean value close to the majority of the data points .. 

## 1. Calc Variance for PMF:

In [68]:
function PmfVar(c)
    pmf = []
    index = []

    for (k, v) in c
        append!(pmf, v/length(data))
        append!(index, k)       
    end
    
    return sum(pmf.*(index - m).^2)
end

PmfVar (generic function with 1 method)

In [69]:
PmfVar(c)

105.55555555555554

Same results as Python codes ..  